src/components/PopupBase/PopupBase.js

```diff
@@ -0,0 +1,172 @@
+    static checkDOMPath(domElem, checkIsTarget) {
+        let currentElem = domElem;
+        while (currentElem && currentElem !== document) {
+            if (checkIsTarget(currentElem)) return true;
+            currentElem = currentElem.parentElement;
+        }
+        return false;
+    }
+
+    constructor(props) {
+        super(props);
+        const { anchorRef, popupRef } = props;
+
+        this.state = { isOpened: false };
+        this.popper = undefined;
+        this.anchorRef = anchorRef || React.createRef();
+        this.popupRef = popupRef || React.createRef();
+
+        this.setPopupVisibility = this.setPopupVisibility.bind(this);
+        this.handleWindowClick = this.handleWindowClick.bind(this);
+        this.handleKeyPress = this.handleKeyPress.bind(this);
+    }
+
+    componentDidMount() {
+        document.addEventListener('click', this.handleWindowClick);
+        document.addEventListener('keydown', this.handleKeyPress, false);
+    }
+
+    componentDidUpdate(prevProps, prevState) {
+        const { isOpened } = this.state;
+        if (isOpened && !prevState.isOpened) {
+            this.createPopperInstance();
+        }
+        if (!isOpened && prevState.isOpened) {
+            this.destroyPopperIstance();
+        }
+    }
+
+    componentWillUnmount() {
+        document.removeEventListener('click', this.handleWindowClick);
+        document.removeEventListener('keydown', this.handleKeyPress, false);
+    }
+
+    setPopupVisibility(shouldBeOpen) {
+        const { isOpened } = this.state;
+        if (shouldBeOpen !== isOpened) {
+            this.setState({ isOpened: shouldBeOpen });
+        }
+    }
+
+    getArgs() {
+        const { isOpened } = this.state;
+        return {
+            setPopupVisibility: this.setPopupVisibility,
+            isOpened,
+        };
+    }
+
+    handleWindowClick(event) {
+        if (!this.popupRef.current) return;
+        const wasPopupClicked = PopupBase.checkDOMPath(
+            event.target,
+            elem => elem.dataset && !!elem.dataset.popup
+        );
+        const wasOutsideClicked = !this.popupRef.current.contains(event.target);
+
+        const wasAnchorClicked =
+            this.anchorRef.current && this.anchorRef.current.contains(event.target);
+        if (!wasPopupClicked && wasOutsideClicked && !wasAnchorClicked) {
+            this.close();
+        }
+    }
+
+    handleKeyPress(event) {
+        if (event.key === ESCAPE_KEY) {
+            this.close();
+        }
+    }
+
+    close() {
+        const { isOpened } = this.state;
+        if (isOpened === true) {
+            this.setState({ isOpened: false });
+        }
+    }
+
+    createPopperInstance() {
+        if (!this.anchorRef.current || !this.popupRef.current) return;
+        const { placement } = this.props;
+        this.destroyPopperIstance();
+        this.popper = new PopperJS(this.anchorRef.current, this.popupRef.current, {
+            placement,
+        });
+    }
```

Comment:
I think it would be better to refactor it to
```
if (this.anchorRef.current && this.popupRef.current) {
        const { placement } = this.props;
        this.destroyPopperIstance();
        this.popper = new PopperJS(this.anchorRef.current, this.popupRef.current, {
            placement,
        });
}
```
just to remove an unnecessary exit point